### ISSUE TITLE
DatumThese?

### ISSUE THREAD

**Issue by SamProtas** [nullpub/datum#14]
From the docs:

```Either isn’t the only type capable of representing errors. Occasionally, computations can partially fail. For these cases, Datum will eventually also wrap the These adt.```

I've partly implemented DatumThese in two projects now (implementing informally the subset of functionality I needed). 

Is DatumThese something you're interested in supporting in this lib? Do you have plans to implement this? Would you accept a PR?

**Comment by baetheus** [comment]
This looks great! I'll give it a once over tonight. Ideally, we'd also have instance factories like [These from fp-ts](https://github.com/gcanti/fp-ts/blob/master/src/These.ts#L141) such as getShow, getEq, getSemigroup, getApply, getChain, and getMonad. If you don't want to implement these I'm willing to go with your PR as is and I'll come up with some time soon to add them. Thanks for the contribution!

**Comment by SamProtas** [comment]
Sure I can implement those as well, makes sense.

**Comment by SamProtas** [comment]
I just ran into a bit of a wrinkle writing tests for the "obvious" way of exporting a getApply for DatumThese.

I was planning to use `datumThese.getMonad(<caller-supplied-left-semigroup>)` for this instance factory given Apply is parent of Monad, but I realized this lead to a surprising inconsistency between the Apply instances for Datum and DatumThese.

the Datum tests include these assertions:

```
// For ap
assert.deepStrictEqual(D.datum.ap(D.pending, D.initial), D.initial);

// For chain
const g = () => D.initial;
assert.deepStrictEqual(D.datum.chain(D.pending, g), D.pending);

// Note in the above, for ap, `initial` beats `pending`, even if "second". But `chain` short circuits on `pending`.
```

For the DatumThese tests to pass I'd need:
```
type A = Apply2C<DT.URI, string>

// This does not match Datum! The DT.pending short-circuits as this comes from a Monad instance
assert.deepStrictEqual(A.ap(DT.pending, DT.initial), DT.pending)
```

The exported Datum.datum "mega instance" has chain/ap definitions that disagree with each other.

There's some debate over how big of a problem this is here with respect to Either.getValidation having a Monad instance:
https://github.com/gcanti/fp-ts/issues/1150

The consensus seems to have been "we should not do this because it's confusing, even if not technically unlawful" and that instance factory is now deprecated.

To match the direction of `fp-ts` without an immediate breaking change I suggest this package:

- Create new standalone instances `export const Functor: ...`
- The current Datum `ap` behavior can still be exported in an Apply instance, but it should have a "special" name, as that behavior has no corresponding Chain/Monad instance (like Either's getApplicativeValidation)
- The default `export const Monad` instance should have consistent ap/chain behavior (short-circuiting)
- Deprecate mega-instances (fp-ts is doing this anyway) with a note pointing to the matching standalone instances ("special" Apply if using ap, Monad otherwise)

Thoughts?

**Comment by baetheus** [comment]
Hm, I've been working on hkts in deno for awhile so I've been out of touch with the fp-ts paradigm. I think it's about time for a major release of datum. For now we can cut an RC of 4.0 with DatumThese and the other changes you've recommended.

I think the current Datum ap instance behavior is unappealing (despite the test). A deprecation of that particular instance seems appropriate.

**Comment by SamProtas** [comment]
Okay, well if you're still interested in merging this PR as-is I think that would make sense. I hope to get time to start another branch off master with the suggestions above around standalone instances and the ap/chain deprecation. If the DatumThese gets merged in while thats in progress, that work can include DatumThese.

**Comment by baetheus** [comment]
I've merged the initial DatumEither work. I'll hold off on cutting a release until you've had some time to look at the other instances. I'll set a reminder to revisit next Friday.

**Comment by baetheus** [comment]
Released in 745aff7b9dca8890addaa67687af1f17045a3e35
